doc/man/man1/maskgen.1
.TH statsgen "1" "PPACK Commands" "November 2020" "PPACK Commands"

.SH "NAME"
statsgen \-

.SH "SYNOPSIS"
\fBstatsgen\fR [\fIOPTIONS\fR...]

.SH "DESCRIPTION"
.PP
Generate statistics to generated optimal masks.

.SH "OPTIONS"

.TP "10"

\f3\-w\fP\f3,\fP \f3\-\-wordlist\fP
Wordlist file to analize.
.IP

\f3\-o\fP\f3,\fP \f3\-\-output\fP
This file contains all the generated masks and they occurrence  that meet all requirements that we have entered.
.IP

.TP
\f3\-i \fP\f3, \fP \f3\-\-input\fP\
Input File. This one contains all the requiements for policygen, so this file can be used instead of all the requirement flags.
.IP


.TP
\fP \f3\-\-scs\fP
Simple charset.
.IP

.TP
\fP \f3\-\-acs\fP
Advance charset.
.IP

.TP
\fP \f3\-\-minlength\fP
Minimum masks' length.
.IP

.TP
\fP \f3\-\-maxlength\fP
Maximum masks' length.
.IP



.TP
\f3\-q\fP\f3,\fP \f3\-\-quiet\fP
Ocult the PPACK's banner.
.IP

.TP
\f3\-\-pretty\fP
Print the result in a pretty format.
.IP

.TP
\fP \f3\-\-hiderare\fP
Hide masks with a occurence percent lower than 1%.
.IP


.SH "EXAMPLES"
This examples ilustrate the use of \f3statsgen\fP:
.RS
.PP
.nf
.ft 3
$ statsgen --wordlist john.txt --output john.masks --minlength 3


.ft 1
.fi
.RE
.PP
This example ilustrate how to enter  \f3statsgen\fP requiments using a \f3ini\fP file, because
sometime there are a lot of requirements:
.RS
.PP
.nf
.ft 3
$ statsgen -i filter.ini

.PP
\fIfilter.ini\fR
        [statsgen]
        WORDLIST="weak.txt"
        OUTPUT="weak.masks"
        HIDERARE=true
        QUIET=false
        SCS={"loweralpha, mixalphanum"}
        ACS={"alphaspecial, numspecial"}
        MINLENGTH=5
        MAXLENGTH=8
        
.TP
\f3Note:\fP
.PP
The ini file no necessary will hace an unique section \fIstatsgen\fR instead it can have diferent sections but the \f3statsgen\fP section will be the unique read.


.ft 1
.fi
.RE
.PP
This example ilustrate how to run statsgen with 4 threads\fIopenmp\fR:
.RS
.PP
.nf
.ft 3
# export OMP_NUM_THREADS=4
$ statsgen --wordlist john.txt --output john.masks --minlength 3
$ policygen -i filter.ini


src/maskgen.cpp
/*
 * maskgen - cli interface
 *
 * Depuration - 29 nov 2020
 * maskgen (cli interface) depurated.
 *
 * No tested functions: ALL FUNCTION TESTED
 *
 * No implemented functions:
 *  
 * NO COMPLETED
 * --- implement the input of parameters using a input file ---
 *
 * NO COMPLETED
 * --- Implement a beatiful output (call helper function  in maskgenWrite) ---
 *
 * NO COMPLETED
 * --- Implement mask writing in a file(call helper function  in maskgenWrite) ---
 *
 *
 *
 * Maintainer: glozanoa <glozanoa@uni.pe>
 *
 */

#ifndef __INCLUDE_STD_IOSTREAM_H__
#define __INCLUDE_STD_IOSTREAM_H__
#include <iostream>
#endif //__INCLUDE_STD_IOSTREAM_H__

#ifndef __INCLUDE_STD_VECTOR_H_
#define __INCLUDE_STD_VECTOR_H_
#include<vector>
#endif //__INCLUDE_STD_VECTOR_H_

using namespace std;

#ifndef __INCLUDE_PROGRAM_OPTIONS_H__
#define __INCLUDE_PROGRAM_OPTIONS_H__
#include<boost/program_options.hpp>
#endif //__INCLUDE_PROGRAM_OPTIONS_H__

//using namespace boost;
namespace po = boost::program_options;


#ifndef __INCLUDE_PPACK_H__
#define __INCLUDE_PPACK_H__
#include "../include/ppack.hpp"
#endif // __INCLUDE_PPACK_H__


int main(int argc ,char* argv[])
{
    try
    {
        po::options_description files("Files I/O");
        files.add_options()
            ("statsgen, s", po::value<string>(), "File generated by statsgen.")
            ("output, o", po::value<string>()->default_value("maskgen.hcmasks"), "Ouput File.")
            ("input, i", po::value<string>()->default_value(""), "Input File.");

        po::options_description print("Print");
        print.add_options()
            ("show", po::value<bool>()->implicit_value(true)->default_value(false), "Show generated masks.")
            ("quiet, q", po::value<bool>()->implicit_value(true)->default_value(false), "Quiet printing(Omit PPACK logo).")
            ("pretty", po::value<bool>()->implicit_value(true)->default_value(false), "Pretty output.");

        po::options_description check("Compare mask o file of mask against an already cracked list.");
        check.add_options()
            //("checkmasks", po::value<vector<Mask>>(), "Masks to compare.")
            ("checkmasksfile", po::value<string>()->default_value(""), "File of mask to compare.");


        po::options_description password("Password Structure");
        password.add_options()
            //("scharset, scs", po::value<vector<SCS>>(), "Password's Simple Charset.")
            //("acharset, acs", po::value<vector<ACS>>(), "Password's Advance Charset.")
            ("minlength", po::value<unsigned int>()->default_value(0), "Minimum password length.")
            ("maxlength", po::value<int>()->default_value(-1), "Maximum password length.")
            ("mincomplexity", po::value<unsigned int>()->default_value(0), "Minimum password Complexity.")
            ("maxcomplexity", po::value<int>()->default_value(100), "Maximum password Complexity.");


        po::options_description frequency("Mask Ocurrence");
        frequency.add_options()
            ("minoccurrence", po::value<unsigned int>()->default_value(0), "Minimum of mask occurrences.")
            ("maxoccurrence", po::value<int>()->default_value(-1), "Maximum of mask occurrences.");


        po::options_description maskgen("Generate customized mask for crack passwords");
        maskgen.add_options()
            ("version, v", "PPACK version.")
            ("help, h", "Show help.");
        maskgen.add(files).add(print).add(check).add(password).add(frequency);

        po::variables_map vm;

        store(po::command_line_parser(argc, argv).
                options(maskgen).run(), vm);

        if(vm.count("help")){
            cout << maskgen << endl;
            return 1;
        }

        if(vm.count("version"))
        {
            //cout << "PPACK  version " + ppack::VERSION << endl;
            cout << "PPACK  version 0.1"  << endl;

            return 1;
        }

        mstruct pargs(vm);
        PPACK::maskgen(pargs);

        return 0;
    }
    catch(std::exception& e)
    {
        cout << e.what() << endl;
        return 1;
    }
}
